beacon-chain/rpc/beacon_server.go

```diff
@@ -136,6 +147,113 @@ func (bs *BeaconServer) BlockTree(ctx context.Context, _ *ptypes.Empty) (*pb.Blo
+	var mostVotes uint64
+	var bestVoteHash string
+	for _, vote := range beaconState.Eth1DataVotes {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		// Let dataVotes be the set of Eth1DataVote objects vote in state.eth1_data_votes where:
+		// vote.eth1_data.block_hash is the hash of an eth1.0 block that is:
+		//   (i) part of the canonical chain
+		//   (ii) >= ETH1_FOLLOW_DISTANCE blocks behind the head
+		//   (iii) newer than state.latest_eth1_data.block_data.
+		// vote.eth1_data.deposit_root is the deposit root of the eth1.0 deposit contract
+		// at the block defined by vote.eth1_data.block_hash.
+		eth1Hash := bytesutil.ToBytes32(vote.DepositRoot)
+		// Verify the block from the vote's block hash exists in the eth1.0 chain and fetch its height.
+		blockExists, blockHeight, err := bs.powChainService.BlockExists(ctx, eth1Hash)
+		if err != nil {
+			log.WithError(err).WithField("blockRoot", fmt.Sprintf("%#x", bytesutil.Trunc(eth1Hash[:]))).
+				Debug("Could not verify block with hash in ETH1 chain")
+			continue
+		}
+		if !blockExists {
+			continue
+		}
+
+		isBehindFollowDistance := big.NewInt(0).Sub(currentHeight, big.NewInt(eth1FollowDistance)).Cmp(blockHeight) >= 0
+		isAheadStateLatestEth1Data := blockHeight.Cmp(stateLatestEth1Height) == 1
+		correctDepositCount := depositCount == vote.DepositCount
+		correctDepositRoot := bytes.Equal(vote.DepositRoot, depositRootAtHeight[:])
+		if blockExists && isBehindFollowDistance && isAheadStateLatestEth1Data && correctDepositCount && correctDepositRoot {
+			dataVotes = append(dataVotes, vote)
+			he, err := ssz.HashedEncoding(reflect.ValueOf(vote))
+			if err != nil {
+				return nil, fmt.Errorf("could not get encoded hash of eth1data object: %v", err)
+			}
+			bestVoteHash, bestVoteHeight, mostVotes, voteCountMap = bs.countVote(bestVoteHash, bestVoteHeight, blockHeight, err, he, mostVotes, vote, voteCountMap)
+		}
+	}
+	// Now we handle the following two scenarios:
+	// If dataVotes is empty:
+	// Let block_hash be the block hash of the ETH1_FOLLOW_DISTANCE'th ancestor of the head of
+	// the canonical eth1.0 chain.
+	// Let deposit_root be the deposit root of the eth1.0 deposit contract in the
+	// post-state of the block referenced by block_hash.
+	if len(dataVotes) == 0 {
+		return bs.defaultEth1DataResponse(ctx, currentHeight, eth1FollowDistance)
+	}
+	// if D is nonempty:
+	// return the best vote
+	v, ok := voteCountMap[bestVoteHash]
+	if ok {
+		return v.eth1Data, nil
+	}
+	return nil, nil
+}
+
+// let best_vote_data be the eth1_data member of D that has the highest vote count (D.count(eth1_data)),
+// breaking ties by favoring block hashes with higher associated block height.
+func (bs *BeaconServer) countVote(bestVoteHash string, bestVoteHeight *big.Int, blockHeight *big.Int, err error, he [32]byte, mostVotes uint64, vote *pbp2p.Eth1Data, voteCountMap map[string]voteHierarchy) (string, *big.Int, uint64, map[string]voteHierarchy) {
```

Comment:
This doesn't need to be a pointer receiver. It can be a standalone method and ideally moved out of this package into some core processing package. A new or existing package under prysm/beacon-chain/core maybe?

Reply:
i moved it but just simply.
if we want to move the creation of the map and protect it from other writers.
it will need more work
wdyt?